instrumentation/logback/logback-appender-1.0/library/src/main/java/io/opentelemetry/instrumentation/logback/appender/v1_0/internal/LoggingEventMapper.java

```diff
@@ -278,25 +297,103 @@ private static void captureKeyValuePairAttributes(
-            attributes.put(key, ((Number) value).doubleValue());
-          } else {
-            attributes.put(getAttributeKey(key), value.toString());
-          }
+        captureKeyValueAttribute(attributes, keyValuePair.key, keyValuePair.value);
+      }
+    }
+  }
+
+  private static void captureKeyValueAttribute(
+      AttributesBuilder attributes, Object key, Object value) {
+    // empty values are not serialized
+    if (key != null && value != null) {
+      String keyStr = key.toString();
+      // preserve type for boolean and numeric values, everything else is converted to String
+      if (value instanceof Boolean) {
+        attributes.put(keyStr, (Boolean) value);
+      } else if (value instanceof Byte
+          || value instanceof Integer
+          || value instanceof Long
+          || value instanceof Short) {
+        attributes.put(keyStr, ((Number) value).longValue());
+      } else if (value instanceof Double || value instanceof Float) {
+        attributes.put(keyStr, ((Number) value).doubleValue());
+      } else if (value.getClass().isArray()) {
+        if (value instanceof boolean[] || value instanceof Boolean[]) {
+          captureKeyArrayValueAttribute(
+              attributes,
+              AttributeKey.booleanArrayKey(keyStr),
+              value,
+              Boolean[]::new,
+              o -> (Boolean) o);
+        } else if (value instanceof byte[]
+            || value instanceof Byte[]
+            || value instanceof int[]
+            || value instanceof Integer[]
+            || value instanceof long[]
+            || value instanceof Long[]
+            || value instanceof short[]
+            || value instanceof Short[]) {
+          captureKeyArrayValueAttribute(
+              attributes,
+              AttributeKey.longArrayKey(keyStr),
+              value,
+              Long[]::new,
+              o -> ((Number) o).longValue());
+        } else if (value instanceof float[]
+            || value instanceof Float[]
+            || value instanceof double[]
+            || value instanceof Double[]) {
+          captureKeyArrayValueAttribute(
+              attributes,
+              AttributeKey.doubleArrayKey(keyStr),
+              value,
+              Double[]::new,
+              o -> ((Number) o).doubleValue());
+        } else {
+          captureKeyArrayValueAttribute(
+              attributes,
+              AttributeKey.stringArrayKey(keyStr),
+              value,
+              String[]::new,
+              String::valueOf);
         }
+      } else if (value instanceof Collection) {
+        captureKeyArrayValueAttribute(
+            attributes,
+            AttributeKey.stringArrayKey(keyStr),
+            ((Collection<?>) value).toArray(),
+            String[]::new,
+            String::valueOf);
+      } else {
+        attributes.put(getAttributeKey(keyStr), String.valueOf(value));
+      }
+    }
+  }
+
+  private static <T> void captureKeyArrayValueAttribute(
```

Comment:
I think it would be easier to write the converted values to a list.
```java
  private static <T> void captureKeyArrayValueAttribute(
      AttributesBuilder attributes,
      AttributeKey<List<T>> key,
      Object array,
      Function<Object, T> extractor) {
    List<T> list = new ArrayList<>();
    int length = java.lang.reflect.Array.getLength(array);
    for (int i = 0; i < length; i++) {
      Object value = (T) java.lang.reflect.Array.get(array, i);
      if (value != null) {
        list.add(extractor.apply(value));
      }
    }
    // empty lists are not serialized
    if (!list.isEmpty()) {
      attributes.put(key, list);
    }
  }
```